drivers/pwm/pwm_led_esp32.c

```diff
@@ -0,0 +1,470 @@
+/*
+ * Copyright (c) 2017 Vitor Massaru Iha <vitor@massaru.org>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Include esp-idf headers first to avoid redefining BIT() macro */
+#include <esp_intr_alloc.h>
+#include <soc/dport_reg.h>
+#include <rom/gpio.h>
+#include <soc/gpio_sig_map.h>
+#include <soc/ledc_reg.h>
+
+#include <soc.h>
+#include <errno.h>
+#include <board.h>
+#include <pwm.h>
+#include <kernel.h>
+#include <gpio.h>
+#include <string.h>
+
+enum {
+	PWM_LED_ESP32_REF_TICK_FREQ,
+	PWM_LED_ESP32_APB_CLK_FREQ,
+};
+
+enum {
+	PWM_LED_ESP32_HIGH_SPEED,
+	PWM_LED_ESP32_LOW_SPEED
+};
+
+#define PWM_LED_ESP32_CHANNEL_NOT_USED -1
+
+/* XXX Move struct peripheral for other place ? */
+struct pwm_led_esp32_peripheral {
+	int clk;
+	int rst;
+};
+
+struct pwm_led_esp32_timer {
+	/* set timer counter bit number [10; 15] */
+	int bit_num;
+	int timer_sel;
+	int freq;
+};
+
+struct pwm_led_esp32_data {
+	/* pin-channel mapping */
+	int pin_ch[40];
```

Comment:
Can't this be a more compact representation? I see it can have 3 different values only, so a `u8_t pin_ch[10]` should suffice if only two bits per I/O pin is used; this should free up a few hundred bytes of RAM.

Reply:
I agree. I'll fix this.

Reply:
I'm adding options related to the pin/channel on Kconfig, and this struct will no longer be needed.

Reply:
Actually I'm using  `{u8_t timer: 2; u8_gpio: 6}` (`pwm_led_esp32_channel ch_cfg[16]`). This is more than you suggested, but now I'm using more options, and I removed the gpio range values.